receiver/couchbasereceiver/client.go

```diff
@@ -0,0 +1,151 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package couchbasereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/couchbasereceiver"
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"go.opentelemetry.io/collector/component"
+	"go.uber.org/zap"
+)
+
+// clusterPath is the path to cluster endpoint
+const clusterPath = "/pools/default"
+
+type client interface {
+	// GetClusterDetails calls "/pools/default" endpoint to get the cluster details
+	GetClusterDetails(ctx context.Context) (*clusterResponse, error)
+
+	// GetBuckets calls the specified bucket path to return details on the buckets associated with the cluster.
+	// The path should be extracted from the clusterResponse.BucketsInfo.URI
+	GetBuckets(ctx context.Context, path string) ([]*bucket, error)
+
+	// GetBucketStats retrieves the stats for the bucket at the given path.
+	// The path should be extraced from bucket.StatsInfo.URI
+	GetBucketStats(ctx context.Context, path string) (*bucketStats, error)
+}
+
+var _ client = (*couchbaseClient)(nil)
+
+type couchbaseClient struct {
+	client       *http.Client
+	hostEndpoint string
+	creds        couchbaseCredentials
+	logger       *zap.Logger
+}
+
+type couchbaseCredentials struct {
+	username string
+	password string
+}
+
+func newClient(cfg *Config, host component.Host, logger *zap.Logger) (client, error) {
+	httpClient, err := cfg.ToClient(host.GetExtensions())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create HTTP Client: %w", err)
+	}
+
+	return &couchbaseClient{
+		client:       httpClient,
+		hostEndpoint: cfg.Endpoint,
+		creds: couchbaseCredentials{
+			username: cfg.Username,
+			password: cfg.Password,
+		},
+		logger: logger,
+	}, nil
+}
+
+func (c *couchbaseClient) GetClusterDetails(ctx context.Context) (*clusterResponse, error) {
+	var clusterInfo clusterResponse
+
+	if err := c.get(ctx, clusterPath, &clusterInfo); err != nil {
+		c.logger.Error("Failed to retrieve cluster data", zap.Error(err))
```

Comment:
This would generate duplicate entries in the log, wouldn't it? The call chain would eventually log the error, and this code here is logging it as well.

Reply:
Yeah, I think you're right. Things that call `get` should likely not log the error. I think the message is useful to indicate what instance of the `get` call failed but likely not also logging the error.

If you agree I'll remove the `zap.Error` fields from the top level API calls.

Reply:
I would remove all c.logger.Error calls that are followed by "return err". Or change them to debug, so that people will get more details only when the appropriate log-level is set.

Reply:
I'll change them to debug, if nothing else if a bug gets filed against these it'll help us figure out what's going wrong.

Reply:
@jpkrohling I changed all the logs in that file to debug logs. I think the error messages capture what is happening and the additional debug logging can help a user or us pinpoint what's going on.